cluster-autoscaler/cloudprovider/hetzner/hetzner_node_group.go

```diff
@@ -0,0 +1,410 @@
+	}
+
+	node := apiv1.Node{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   newNodeName(n),
+			Labels: map[string]string{},
+		},
+		Status: apiv1.NodeStatus{
+			Capacity:   resourceList,
+			Conditions: cloudprovider.BuildReadyConditions(),
+		},
+	}
+	node.Status.Allocatable = node.Status.Capacity
+	node.Labels = cloudprovider.JoinStringMaps(node.Labels, buildNodeGroupLabels(n))
+	node.Status.Conditions = cloudprovider.BuildReadyConditions()
+
+	nodeInfo := schedulerframework.NewNodeInfo(cloudprovider.BuildKubeProxy(n.id))
+	err = nodeInfo.SetNode(&node)
+	if err != nil {
+		return nil, fmt.Errorf("could not create node info for node group %s error: %v", n.id, err)
+	}
+
+	return nodeInfo, nil
+}
+
+// Exist checks if the node group really exists on the cloud provider side.
+// Allows to tell the theoretical node group from the real one. Implementation
+// required.
+func (n *hetznerNodeGroup) Exist() bool {
+	_, exists := n.manager.nodeGroups[n.id]
+	return exists
+}
+
+// Create creates the node group on the cloud provider side. Implementation
+// optional.
+func (n *hetznerNodeGroup) Create() (cloudprovider.NodeGroup, error) {
+	n.manager.nodeGroups[n.id] = n
+
+	return n, cloudprovider.ErrNotImplemented
+}
+
+// Delete deletes the node group on the cloud provider side.  This will be
+// executed only for autoprovisioned node groups, once their size drops to 0.
+// Implementation optional.
+func (n *hetznerNodeGroup) Delete() error {
+	// We do not use actual node groups but all nodes within the Hcloud project are labeled with a group
+	return nil
+}
+
+// Autoprovisioned returns true if the node group is autoprovisioned. An
+// autoprovisioned group was created by CA and can be deleted when scaled to 0.
+func (n *hetznerNodeGroup) Autoprovisioned() bool {
+	// All groups are auto provisioned
+	return false
+}
+
+func toInstance(vm *hcloud.Server) cloudprovider.Instance {
+	return cloudprovider.Instance{
+		Id:     toProviderID(vm.ID),
+		Status: toInstanceStatus(vm.Status),
+	}
+}
+
+func toProviderID(nodeID int) string {
+	return fmt.Sprintf("%s%d", providerIDPrefix, nodeID)
+}
+
+func toInstanceStatus(status hcloud.ServerStatus) *cloudprovider.InstanceStatus {
+	if status == "" {
+		return nil
+	}
+
+	st := &cloudprovider.InstanceStatus{}
+	switch status {
+	case hcloud.ServerStatusInitializing:
+	case hcloud.ServerStatusStarting:
+		st.State = cloudprovider.InstanceCreating
+	case hcloud.ServerStatusRunning:
+		st.State = cloudprovider.InstanceRunning
+	case hcloud.ServerStatusOff:
+	case hcloud.ServerStatusDeleting:
+	case hcloud.ServerStatusStopping:
+		st.State = cloudprovider.InstanceDeleting
+	default:
+		st.ErrorInfo = &cloudprovider.InstanceErrorInfo{
+			ErrorClass:   cloudprovider.OtherErrorClass,
+			ErrorCode:    "no-code-hcloud",
+			ErrorMessage: "error",
+		}
+	}
+
+	return st
+}
+
+func newNodeName(n *hetznerNodeGroup) string {
+	return fmt.Sprintf("%s-%d", n.id, rand.Int63())
```

Comment:
We could consider including the cluster name in the node name as well, which can be specified through the `--cluster-name` parameter.
On the other hand, it could be considered "best practice" to have only one cluster oer Hetzner Cloud project.
Just food for thought, I don't think it's a necessary change.

Reply:
Using one cluster per project has some downsides, given that authentication has to be setup manually per project: There are no (documented) API calls to setup authentication tokens for a new project.

Reply:
What @WebSpider said, we can not create a new project and api key so to ensure we need as little manual intervention as possible I decided to go for node labels.